src/libsyntax_ext/deriving/custom.rs

```diff
@@ -12,66 +12,76 @@ use std::panic;
 
 use errors::FatalError;
 use proc_macro::{TokenStream, __internal};
-use syntax::ast::{self, ItemKind};
-use syntax::codemap::{ExpnInfo, MacroAttribute, NameAndSpan, Span};
+use syntax::ast::{self, ItemKind, MetaItemKind, Attribute};
+use syntax::attr::mark_used;
+use syntax::codemap::Span;
 use syntax::ext::base::*;
-use syntax::fold::Folder;
-use syntax::parse::token::intern;
-use syntax::print::pprust;
+use syntax::parse::token::InternedString;
+
+fn mark_attrs_used<I: Iterator<Item=Attribute>>(attrs: I, whitelist: &[InternedString]) {
+    for attr in attrs {
+        let name = match attr.node.value.node {
+            MetaItemKind::Word(ref n) => n,
+            MetaItemKind::List(ref n, _) => n,
+            MetaItemKind::NameValue(ref n, _) => n,
+        };
+
+        if whitelist.contains(name) {
+            mark_used(&attr);
+        }
+    }
+}
 
 pub struct CustomDerive {
     inner: fn(TokenStream) -> TokenStream,
+    attrs: Vec<InternedString>,
 }
 
 impl CustomDerive {
-    pub fn new(inner: fn(TokenStream) -> TokenStream) -> CustomDerive {
-        CustomDerive { inner: inner }
+    pub fn new(inner: fn(TokenStream) -> TokenStream,
+               attrs: &[InternedString])
+               -> CustomDerive {
+        CustomDerive { inner: inner, attrs: attrs.into() }
     }
 }
 
-impl MultiItemModifier for CustomDerive {
+impl MultiItemDecorator for CustomDerive {
     fn expand(&self,
               ecx: &mut ExtCtxt,
               span: Span,
-              meta_item: &ast::MetaItem,
-              item: Annotatable)
-              -> Vec<Annotatable> {
-        let item = match item {
-            Annotatable::Item(item) => item,
-            Annotatable::ImplItem(_) |
-            Annotatable::TraitItem(_) => {
+              _meta_item: &ast::MetaItem,
+              item: &Annotatable,
+              push: &mut FnMut(Annotatable)) {
+        let item_ref = match item {
+            &Annotatable::Item(ref item) => item,
+            &Annotatable::ImplItem(_) |
+            &Annotatable::TraitItem(_) => {
                 ecx.span_err(span, "custom derive attributes may only be \
                                     applied to struct/enum items");
-                return Vec::new()
+                return;
             }
         };
-        match item.node {
-            ItemKind::Struct(..) |
-            ItemKind::Enum(..) => {}
+        mark_attrs_used(item_ref.attrs.iter().cloned(), &self.attrs);
+        match item_ref.node {
+            ItemKind::Struct(ref var, ..) =>
+                mark_attrs_used(var.fields().iter().flat_map(|f| f.attrs.iter()).cloned(),
+                                &self.attrs),
+            ItemKind::Enum(ref def, ..) =>
+                mark_attrs_used(def.variants.iter().flat_map(|f| f.node.attrs.iter()).cloned(),
+                                &self.attrs),
```

Comment:
If we use a visitor, these call should be able to be replaced with `MarkUsed(&self.attrs).visit_item(item_ref)`.